### PR TITLE
Better error message for determinants of PSDMatrix types with non-square factor

### LINKED ISSUE #16: Better error message for PSD matrix defined by non-square factor

### PR DESCRIPTION
**Pull request by pnkraemer**
Resolves #16.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -10,7 +10,8 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 julia = "1.6,1.7"
 
 [extras]
+Suppressor = "fd094767-a336-5f1f-9728-57cf17d0bbfb"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [targets]
-test = ["Test"]
+test = ["Test", "Suppressor"]
```
Comment (pnkraemer) [inline Project.toml]: please let me know if this is the incorrect way of specifying test dependencies!

File: test/runtests.jl [hunk 2]
```diff
@@ -24,17 +25,17 @@ eltypes = (Int64, Float64, BigFloat)
             if size(M, 1) == size(M, 2)
                 @test todense(inv(S)) ≈ inv(todense(S))
             end
-            @test show(S) == nothing
-            @test show(stdout, MIME("text/plain"), S) == nothing
+            @suppress_out @test isnothing(show(S))
+            @suppress_out @test isnothing(show(stdout, MIME("text/plain"), S))
```
Comment (pnkraemer) [inline test/runtests.jl:31]: (i) I found it a bit annoying to always have the prints, so I suppressed it (which, unfortunately, introduces an additional dependency...)
(ii) I'll be honest, VSCode complained about `== nothing`, so then I googled and found out that there is `isnothing()` (https://github.com/JuliaLang/julia/issues/29674). Please let me know if this change is not welcome!

File: Project.toml (hunk 1 above)
Comment (nathanaelbosch) [reply Project.toml]: Actually yes, test dependencies are supposed to be somewhere else: https://github.com/nathanaelbosch/ProbNumDiffEq.jl/blob/main/test/Project.toml

TL;DR:
```
]activate test
add Suppressor
```

File: test/runtests.jl (hunk 2 above)
Comment (nathanaelbosch) [reply test/runtests.jl:31]: True, `== nothing` is discouraged, my bad. And thanks for changing it to `isnothing()` :)

File: src/PSDMatrices.jl [hunk 3]
```diff
@@ -30,8 +30,27 @@ end
...
+
+function logdet(M::PSDMatrix)
+    if size(M.R, 1) != size(M.R, 2)
+        msg = _errormsg_non_square_factor(M)
+        throw(DimensionMismatch(msg))
+    end
+    return 2 * logdet(M.R)
+end
+
+_errormsg_non_square_factor(M::PSDMatrix) = (
```
Comment (nathanaelbosch) [inline src/PSDMatrices.jl]: Overall this looks fine, but I'm wondering about some details here:
1. did you consider making the function `_non_square_factor_error`, that returns a `DimensionMismatch` with the correct error?
2. or alternatively, have a custom exception? (but I'm not sure how that would be called)
3. and on a different level: Do we need the leading underscore? I'm not sure about Julia conventions here. But since it's not exported, I'd assume that it should also be safe to give it a normal function name. What do you think?

**Review by nathanaelbosch — CHANGES_REQUESTED** [review]

File: src/PSDMatrices.jl (hunk 3 above)
Comment (pnkraemer) [reply src/PSDMatrices.jl]: 1. Yes, but I personally find it slightly more readable if the error is raised on top-level (i.e. on the same level as the return). But this is subjective, and I can change it if you like. 
2. I briefly thought about this, but I don't know which exception that would be... and it might be a bad idea to start construction custom exceptions left and right? Alternatively, one could raise a `MethodError` here, because strictly speaking, the error is not that the dimensions don't match but that PSDMatrices does not know how to do this efficiently. Thoughts?
3. I was wondering the same and found https://stackoverflow.com/questions/53687998/function-name-with-a-leading-underscore. I understood this to be a mild stylistic convention, but it looks familiar to me so I liked it :D Shall I change it?

File: Project.toml (hunk 1 above)
Comment (pnkraemer) [reply Project.toml]: Thanks! I updated it.

File: src/PSDMatrices.jl (hunk 3 above)
Comment (nathanaelbosch) [reply src/PSDMatrices.jl]: 1. no strong preference so let's keep it this way :+1: 
2. now that you mention it, maybe I'd slightly favor a `MethodError` here, but again no real preferences. Regarding a custom error, I again have no idea about conventions, but since I don't know what exactly it should be called I agree that we should rather use existing ones.
3. I also found that stackoverflow question, but nothing else really. If you don't have a preference I'd tend towards not using a leading underscore here to mark the function as internal, since unless exported they all are.

Comment (nathanaelbosch) [reply src/PSDMatrices.jl]: But to bring another wild suggestion to the table: What about something like
```julia
function ensure_square_factor(M::PSDMatrix)
    msg = (
        "The requested operation is not available for a PSDMatrix with a non-square factor." *
        "The factor of the received PSDMatrix has dimensions ($(size(M.R,1)), $(size(M.R,2))). " *
        "Try turning the PSDMatrix into a dense matrix first."
    )
    if size(M.R, 1) != size(M.R, 2)
        throw(DimensionMismatch(msg))
    end
end
```
? Maybe we need to check for square square roots (lol) in more places, so that deduplicates things a bit? 

But again, just a random suggestion and if you prefer the current approach that's fine with me :)

Comment (pnkraemer) [reply src/PSDMatrices.jl]: I like that idea and implemented it! The only change i made is to use `confirm` instead of `ensure`, because it sounds more like an `if not, raise error` check and less like an `if not, transform` check.

Comment (nathanaelbosch) [reply src/PSDMatrices.jl]: Nice :D I also wasn't completely happy with `ensure`, `confirm` is better indeed! :+1:

**Review by nathanaelbosch — APPROVED** [review]
Thanks for the PR! :)
